### PR TITLE
OVS/GRE: bug fixes

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes some issues with OVS/GRE
Issues with isolated networks/vpcs: #2886 #2996  #3802
Issues with shared networks: #2863 #2885

To test this PR, we need to create a zone with isolation method 'GRE' instead of default 'VLAN'.

Test results
(1) smoke test test/integration/smoke/test_privategw_acl_ovs_gre.py passed on kvm and xenserver
(2) component test test/integration/component/test_vpc_distributed_routing_offering.py passed on xenserver.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xenserver-71 keepEnv

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xenserver-71 keepEnv

**Comment by svenvogel** [comment]
@weizhouapache only for interest because we are using OVS/VLAN at the moment with KVM. What are the fixes in detail? 😄 thanks and cheers

**Comment by weizhouapache** [comment]
> @weizhouapache only for interest because we are using OVS/VLAN at the moment with KVM. What are the fixes in detail?  thanks and cheers

@svenvogel 
this PR fixes some issues with Openvswitch/GRE support on kvm and xenserver. 
You can find the description of issues in each commit.

OVS/VLAN is not impacted.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by rhtyd** [comment]
@weizhouapache is this critical for 4.16, and is this ready for review - how are we testing it?

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
> @weizhouapache is this critical for 4.16, and is this ready for review - how are we testing it?

@rhtyd it is ready for review.
if we want to support OVS/GRE in 4.16.0.0 and future versions, this is critical (even blocker as the feature is totally broken on kvm, and not working well on xenserver).

To test this PR, we need to create a zone with isolation method 'GRE' instead of default 'VLAN'.

(for existing zone, there is no api to update the isolation method. we can change it via DB and cmk, for example
MySQL> update physical_network_isolation_methods set isolation_method ='GRE' where id=2 and physical_network_id=201;
cmk> add networkserviceprovider physicalnetworkid=bbb96664-cc84-4a66-af66-fd79b8fdf59a name=Ovs
cmk> update networkserviceprovider state=Enabled id=7b806b85-bcfb-4876-8da1-1644dc229a06
)

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@rhtyd @weizhouapache I don't think this is a blocker as it has been broken for many releases, however, it would be great to have it on 4.16.0

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtOvsFetchInterfaceCommandWrapper.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 package com.cloud.hypervisor.kvm.resource.wrapper;
 
+import com.google.common.base.Strings;
```
Comment (GutoVeronezi) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtOvsFetchInterfaceCommandWrapper.java:22]: Recently we finished a voting defining `commons.lang3` as the default string lib ([VOTE thread](https://lists.apache.org/thread.html/r806cd10b3de645c150e5e0e3d845c5a380a700197143f57f0834d758%40%3Cdev.cloudstack.apache.org%3E) & [RESULT thread](https://lists.apache.org/thread.html/r5898f717d506a0e160b9dc766b65769e2fe2f07f9cba70825843e272%40%3Cdev.cloudstack.apache.org%3E)). The code and the docs will be changed soon, could you already use `commons.lang3` here?

**Comment by weizhouapache** [comment]
> @rhtyd @weizhouapache I don't think this is a blocker as it has been broken for many releases, however, it would be great to have it on 4.16.0

agree with @nvazquez 
We can support it again since 4.16.0.0

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtOvsFetchInterfaceCommandWrapper.java (hunk 1 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtOvsFetchInterfaceCommandWrapper.java]: @GutoVeronezi done.

Comment (GutoVeronezi) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtOvsFetchInterfaceCommandWrapper.java]: @weizhouapache thanks!

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@sureshanaparti did you ping the interested parties for testing? Or do you have any advise on testing?

**Comment by weizhouapache** [comment]
> @sureshanaparti did you ping the interested parties for testing? Or do you have any advise on testing?

@sureshanaparti @DaanHoogland @nvazquez 
I think nobody is using this setting now (advanced zone with GRE isolation) as the feature was broken in some previous versions.

My opinion is
(1) if anyone want to test it, it is fine. ping me if you need help.
(2) if nobody have chance testing it, I think we can merge it if we make sure it does not break existing functionalities (by code review, smoke test, manual test, etc).

I have tested the feature manually, 
the new smoke test (test/integration/smoke/test_privategw_acl_ovs_gre.py) passed on kvm and xenserver. component test test_vpc_distributed_routing_offering.py passed on xenserver, but not on kvm as the feature is never implemented on kvm.

if someone find bugs in their testing, we can fix them in next minor releases (4.16 will be LTS as far as I know)

**Review by nvazquez — APPROVED** [review]
Edit: Still testing on OVS-KVM environment

Manually updated database to set GRE isolation on the existing physical network and added and enabled the Ovs provider. 

vpc-test2

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]

**Pull request merged**